File: backend/app/core/url_validation.py
from urllib.parse import parse_qs, urlparse

from backend.app.core.app_errors import AppError, validation_error


def _field_name(index: int) -> str:
    return f"videos[{index}].url"


def _parse_http_url(raw_url: str) -> tuple[str, object | None]:
    url = raw_url.strip()
    if not url:
        return url, None
    parsed = urlparse(url)
    if parsed.scheme.lower() not in {"http", "https"} or not parsed.netloc:
        return url, None
    return url, parsed


def _host_matches(host: str, root: str) -> bool:
    return host == root or host.endswith(f".{root}")


def _has_path_token(path: str, index: int = 0) -> bool:
    parts = [part for part in path.split("/") if part]
    return len(parts) > index and bool(parts[index].strip())


def is_valid_youtube_url(url: str) -> bool:
    _, parsed = _parse_http_url(url)
    if parsed is None:
        return False

    host = (parsed.hostname or "").lower()
    path = parsed.path or ""
    if host == "youtu.be":
        return _has_path_token(path)
    if not _host_matches(host, "youtube.com"):
        return False

    if path.rstrip("/").lower() == "/watch":
        video_ids = parse_qs(parsed.query).get("v") or []
        return any(video_id.strip() for video_id in video_ids)
    parts = [part for part in path.split("/") if part]
    return len(parts) >= 2 and parts[0].lower() == "shorts" and bool(parts[1].strip())


def is_valid_instagram_reel_url(url: str) -> bool:
    _, parsed = _parse_http_url(url)
    if parsed is None:
        return False

    host = (parsed.hostname or "").lower()
    if not _host_matches(host, "instagram.com"):
        return False
    parts = [part for part in (parsed.path or "").split("/") if part]
    return len(parts) >= 2 and parts[0].lower() == "reel" and bool(parts[1].strip())


def validate_video_url(
    platform: str,
    url: str,
    video_id: str | None = None,
    field: str | None = None,
) -> AppError | None:
    normalized_platform = platform.lower().strip()
    trimmed = url.strip()
    if not trimmed:
        return validation_error("Enter a URL for this video.", video_id=video_id, field=field)

    if normalized_platform == "youtube":
        if is_valid_youtube_url(trimmed):
            return None
        if is_valid_instagram_reel_url(trimmed):
            return validation_error(
                "Selected platform is YouTube, but the URL is an Instagram Reel.",
                video_id=video_id,
                field=field,
                code="VALIDATION_PLATFORM_URL_MISMATCH",
            )
        return validation_error(
            "Enter a supported YouTube URL: youtube.com/watch, youtube.com/shorts, or youtu.be.",
            video_id=video_id,
            field=field,
        )

    if normalized_platform == "instagram":
        if is_valid_instagram_reel_url(trimmed):
            return None
        if is_valid_youtube_url(trimmed):
            return validation_error(
                "Selected platform is Instagram, but the URL is a YouTube video.",
                video_id=video_id,
                field=field,
                code="VALIDATION_PLATFORM_URL_MISMATCH",
            )
        return validation_error(
            "Enter a supported Instagram Reel URL in the form instagram.com/reel/...",
            video_id=video_id,
            field=field,
        )

    return validation_error(
        "Unsupported video platform.",
        video_id=video_id,
        field=field,
        code="VALIDATION_UNSUPPORTED_PLATFORM",
    )


def validate_ingest_videos(videos: list[dict]) -> tuple[list[dict], AppError | None]:
    normalized: list[dict] = []
    for index, video in enumerate(videos):
        video_id = str(video.get("video_id") or ("A" if index == 0 else "B"))
        platform = str(video.get("platform") or "").lower().strip()
        url = str(video.get("url") or "").strip()
        error = validate_video_url(platform, url, video_id=video_id, field=_field_name(index))
        if error is not None:
            return [], error
        normalized.append({**video, "video_id": video_id, "platform": platform, "url": url})
    return normalized, None

File: frontend/src/lib/urlValidation.ts
export type Platform = "youtube" | "instagram";

export type VideoInputForValidation = {
  video_id: "A" | "B";
  platform: Platform;
  url: string;
};

export type VideoValidation = {
  normalizedUrl: string;
  error: string | null;
};

function parseHttpUrl(rawUrl: string): URL | null {
  const trimmed = rawUrl.trim();
  if (!trimmed) return null;
  try {
    const url = new URL(trimmed);
    if (url.protocol !== "http:" && url.protocol !== "https:") return null;
    return url;
  } catch {
    return null;
  }
}

function hostMatches(hostname: string, root: string) {
  const host = hostname.toLowerCase();
  return host === root || host.endsWith(`.${root}`);
}

function hasPathToken(pathname: string, index = 0) {
  const parts = pathname.split("/").filter(Boolean);
  return parts.length > index && Boolean(parts[index]?.trim());
}

export function isValidYouTubeUrl(rawUrl: string) {
  const url = parseHttpUrl(rawUrl);
  if (!url) return false;

  const host = url.hostname.toLowerCase();
  if (host === "youtu.be") return hasPathToken(url.pathname);
  if (!hostMatches(host, "youtube.com")) return false;

  if (url.pathname.replace(/\/+$/, "").toLowerCase() === "/watch") {
    return Boolean(url.searchParams.get("v")?.trim());
  }

  const parts = url.pathname.split("/").filter(Boolean);
  return parts.length >= 2 && parts[0]?.toLowerCase() === "shorts" && Boolean(parts[1]?.trim());
}

export function isValidInstagramReelUrl(rawUrl: string) {
  const url = parseHttpUrl(rawUrl);
  if (!url || !hostMatches(url.hostname, "instagram.com")) return false;

  const parts = url.pathname.split("/").filter(Boolean);
  return parts.length >= 2 && parts[0]?.toLowerCase() === "reel" && Boolean(parts[1]?.trim());
}

export function validateVideoUrl(platform: Platform, rawUrl: string): VideoValidation {
  const normalizedUrl = rawUrl.trim();
  if (!normalizedUrl) {
    return { normalizedUrl, error: "Enter a URL for this video." };
  }

  if (platform === "youtube") {
    if (isValidYouTubeUrl(normalizedUrl)) return { normalizedUrl, error: null };
    if (isValidInstagramReelUrl(normalizedUrl)) {
      return { normalizedUrl, error: "Selected platform is YouTube, but the URL is an Instagram Reel." };
    }
    return {
      normalizedUrl,
      error: "Enter a supported YouTube URL: youtube.com/watch, youtube.com/shorts, or youtu.be."
    };
  }

  if (isValidInstagramReelUrl(normalizedUrl)) return { normalizedUrl, error: null };
  if (isValidYouTubeUrl(normalizedUrl)) {
    return { normalizedUrl, error: "Selected platform is Instagram, but the URL is a YouTube video." };
  }
  return { normalizedUrl, error: "Enter a supported Instagram Reel URL in the form instagram.com/reel/..." };
}

export function normalizeVideoInputs(videos: VideoInputForValidation[]) {
  return videos.map((video) => ({
    ...video,
    platform: video.platform,
    url: video.url.trim()
  }));
}

export function duplicateUrlWarning(videos: VideoInputForValidation[]) {
  const [first, second] = videos.map((video) => video.url.trim()).filter(Boolean);
  if (first && second && first === second) {
    return "Video A and Video B use the same URL. Submission is allowed.";
  }
  return null;
}

export function sameInputs(a: VideoInputForValidation[] | null, b: VideoInputForValidation[] | null) {
  if (!a || !b || a.length !== b.length) return false;
  return a.every((video, index) => {
    const other = b[index];
    return (
      other &&
      video.video_id === other.video_id &&
      video.platform === other.platform &&
      video.url.trim() === other.url.trim()
    );
  });
}
